### PR TITLE
Fix stats to not include queued time in planning time

### LINKED ISSUE #5980: Presto stats planning time is including queueing time
I noticed this when Presto was reporting super long planning times, but it looks like this was just because the queues are really full.

### PR DESCRIPTION
**Pull request by idemura**
Fixes #5980

### CODE REVIEW

**Comment by haozhun** [comment]
Please rebase your change and remove the merge records in your commit history. Please fix the checkstyle failures as well.

**Comment by idemura** [comment]
Done all.

**Comment by haozhun** [comment]
See http://chris.beams.io/posts/git-commit/ for commit message guidelines. Describe what you changed in the commit message. Don't use the name of the issue.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 1]
```diff
@@ -472,14 +473,14 @@ public boolean isDone()
...
         return queryState.compareAndSet(QUEUED, PLANNING);
     }
 
     public boolean transitionToStarting()
     {
-        Duration durationSinceCreation = nanosSince(createNanos).convertToMostSuccinctTimeUnit();
-        queuedTime.compareAndSet(null, durationSinceCreation);
-        totalPlanningTime.compareAndSet(null, durationSinceCreation);
+        queuedTime.compareAndSet(null, nanosSince(createNanos).convertToMostSuccinctTimeUnit());
+        recordPlanningTime();
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: I can see pros and cons for inlining vs extracting this method. I think inlining is a better idea given the simplicity of the method and the repetitive nature of other code around this line.

**Comment by haozhun** [comment]
Let's add an additional factory to `QueryStateMachine` (another `begin` method) that takes guava `Ticker`. And change the rest of `QueryStateMachine` accordingly by having its own `nanosSince` method that respects the Ticker and replace all `nanoTime` calls.

Once you do that, you can improve the tests to remove `sleep`.

Make sure the `Ticker` change goes in a separate commit. You have two ways, both acceptable:
* Add the Ticker change in a commit before your current commit
* Split your current commit into two (non-test and test), and add the Ticker commit in the middle.

You can choose whichever one you find more convenient.

**Comment by idemura** [comment]
I'm glad that you like my dependency injection idea.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 1 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: This code was duplicated and it took me failed test to realize that there are several occurrences. For sakes for maintainability, let us have it in one place. Focused named method never a bad idea - especially, it self documents with name record planning time). Probably offline.

**Comment by idemura** [comment]
Changed comment description.

**Comment by haozhun** [comment]
@idemura  Sorry for my change of mind. Kudos to you for proposing this. 👍

**Comment by haozhun** [comment]
I would use commit message: Fix stats to not include queued time in planning time

**Comment by idemura** [comment]
Updated. Dependency injected.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 2]
```diff
@@ -693,4 +723,14 @@ public void pruneQueryInfo()
         );
         finalQueryInfo.compareAndSet(finalInfo, Optional.of(prunedQueryInfo));
     }
+
+    private long tickerNanos()
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: This method is simple enough that I don't see much value in making it a function. I would inline it. Let me know if you see particular benefits in abstracting it here.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 3]
```diff
@@ -166,7 +183,20 @@ public static QueryStateMachine begin(QueryId queryId, String query, Session ses
      */
     public static QueryStateMachine failed(QueryId queryId, String query, Session session, URI self, TransactionManager transactionManager, Executor executor, Throwable throwable)
     {
-        QueryStateMachine queryStateMachine = new QueryStateMachine(queryId, query, session, self, false, transactionManager, executor);
+        return failedWithTicker(queryId, query, session, self, transactionManager, executor, Ticker.systemTicker(), throwable);
+    }
+
+    public static QueryStateMachine failedWithTicker(
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Make this method package-private.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java [hunk 4]
```diff
@@ -137,6 +141,19 @@ private QueryStateMachine(QueryId queryId, String query, Session session, URI se
      */
     public static QueryStateMachine begin(QueryId queryId, String query, Session session, URI self, boolean transactionControl, TransactionManager transactionManager, Executor executor)
     {
+        return beginWithTicker(queryId, query, session, self, transactionControl, transactionManager, executor, Ticker.systemTicker());
+    }
+
+    public static QueryStateMachine beginWithTicker(
```
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Make this method package-private.

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStateMachine.java [hunk 5]
```diff
@@ -58,6 +58,22 @@
 
 public class TestQueryStateMachine
 {
+    private static class MockTicker extends Ticker
+    {
+        @ Override
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStateMachine.java]: I'm actually surprise that having a space after `@` works. Let's remove the space.

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStateMachine.java [hunk 6]
```diff
@@ -58,6 +58,22 @@
...
+        {
+            return timeNanos;
+        }
+
+        public void setTimeMillis(long millis)
+        {
+            timeNanos = millis * 1000000;
+        }
+
+        private long timeNanos;
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStateMachine.java]: We put fields at the top of a class.

File: presto-main/src/test/java/com/facebook/presto/execution/TestQueryStateMachine.java [hunk 7]
```diff
@@ -58,6 +58,22 @@
 
 public class TestQueryStateMachine
 {
+    private static class MockTicker extends Ticker
```
Comment (haozhun) [inline presto-main/src/test/java/com/facebook/presto/execution/TestQueryStateMachine.java]: Use `io.airlift.testing.TestingTicker`.

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 4 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Done

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 3 above)
Comment (idemura) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Done

File: presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java (hunk 2 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/execution/QueryStateMachine.java]: Let's leave this alone.

**Pull request merged**